typed-builder-macro/src/enum_info.rs

```diff
@@ -0,0 +1,158 @@
+use convert_case::{Case, Casing};
+use proc_macro2::{Span, TokenStream};
+use quote::{format_ident, quote, ToTokens};
+use syn::{parse::Error, parse_quote, punctuated::Punctuated, Token};
+
+use crate::builder_attr::{IntoSetting, TypeBuilderAttr};
+use crate::struct_info::StructInfo;
+
+pub struct EnumInfo<'a> {
+    ast: &'a syn::DeriveInput,
+    variants: Vec<&'a syn::Variant>,
+}
+
+impl<'a> EnumInfo<'a> {
+    pub fn new(ast: &'a syn::DeriveInput, variants: impl Iterator<Item = &'a syn::Variant>) -> syn::Result<EnumInfo<'a>> {
+        if !ast.generics.params.is_empty() {
+            return Err(Error::new_spanned(
+                &ast.generics,
+                "TypedBuilder is not supported for enum with generics or lifetime",
+            ));
+        }
+        let builder_attr = TypeBuilderAttr::new(&ast.attrs)?;
+        if builder_attr.builder_method.name.is_some() {
+            return Err(Error::new_spanned(
+                ast,
+                "TypedBuilder is not supported for enum with builder_method(name=...)",
+            ));
+        }
+        if builder_attr.builder_type.name.is_some() {
+            return Err(Error::new_spanned(
+                ast,
+                "TypedBuilder is not supported for enum with builder_type(name=...)",
+            ));
+        }
+        if !matches!(builder_attr.build_method.into, IntoSetting::NoConversion) {
+            return Err(Error::new_spanned(
+                ast,
+                "TypedBuilder is not supported for enum with build_method(into=...)",
+            ));
+        }
+        Ok(EnumInfo {
+            ast,
+            variants: variants.collect(),
+        })
+    }
+
+    fn derive_variant_impl(
+        &self,
+        variant_name: &syn::Ident,
+        variant_attrs: &[syn::Attribute],
+        variant_fields: &syn::FieldsNamed,
+    ) -> syn::Result<TokenStream> {
+        let enum_name = &self.ast.ident;
+        let internal_struct_name = format_ident!("{}{}", enum_name, variant_name);
+        let internal_struct_ast = &syn::DeriveInput {
```

Comment:
I'd rather create a `StructInfo` from the variant than fabricate a `DeriveInput`. Even if it means altering `StructInfo::new` to support this.

Reply:
Creating a `StructInfo` using the constructor is also acceptable here - most of the fields you need to fill are ones you manually fabricate in this `DeriveInput` anyway.